protocols/gossipsub/src/mcache.rs

```diff
@@ -0,0 +1,274 @@
+extern crate fnv;
+
+use fnv::FnvHashMap;
+use libp2p_floodsub::TopicHash;
+use protocol::GossipsubMessage;
+
+/// CacheEntry stored in the history
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct CacheEntry {
+    mid: String,
+    topics: Vec<TopicHash>,
+}
+
+/// MessageCache struct holding history of messages
+#[derive(Debug, Clone, PartialEq)]
+pub struct MessageCache {
+    msgs: FnvHashMap<String, GossipsubMessage>,
```

Comment:
Maybe a hash function which offers better protection against collision attacks should be used here, especially given that message IDs are taken from the network.

Reply:
Yep, I thought I had changed this in an earlier version. Must have escaped this PR. Will update